llvm/lib/CodeGen/MachineLateInstrsCleanup.cpp

```diff
@@ -189,7 +189,13 @@ static bool isCandidate(const MachineInstr *MI, Register &DefedReg,
       if (MO.isDef()) {
         if (i == 0 && !MO.isImplicit() && !MO.isDead())
           DefedReg = MO.getReg();
-        else
+        else if (i != 0 && DefedReg != MCRegister::NoRegister) {
+          if (MO.isDead() && MO.isImplicit())
+            continue;
+          if (MO.isImplicit() && TRI->regsOverlap(MO.getReg(), DefedReg))
```

Comment:
Example:
// If a later implicit def overlaps with the first def, ignore it.
"Overlaps" seem like the wrong condition here. Don't you need to check that its a subset of the first def?

What I'd really like to see is an explanation of _exactly_ what kind of instruction we're looking for, e.g. "must have at least one explicit def and any other def must either be dead or a subset of the first def". And then make sure that the code matches the comment.

Reply:
In the general case I don't think it's safe to remove an instruction like this, because that it removing a def of the upper bits of $rax, which could be used by some other instruction:
```
renamable $eax = MOV32r0 implicit-def dead $eflags, implicit-def $rax
```

Reply:
The implicit operands could be anything, the extra operand could be an arbitrary def of an unrelated register.

I'm assuming this pattern appeared as a result of coalescing SUBREG_TO_REG. We should really try to remove it and require emitting instructions that explicitly define the used register parts. We have to go out our way to maintain the liveness it introduced in the producer instructions, and it introduces a lot of complexity in downstream optimizations (like this case). But that will be a big project, particularly for x86

Reply:
> In the general case I don't think it's safe to remove an instruction like this, because that it removing a def of the upper bits of $rax, which could be used by some other instruction:
> 
> ```
> renamable $eax = MOV32r0 implicit-def dead $eflags, implicit-def $rax
> ```
 
As rax is super register of eax so on changing the value of eax implicitly there will be change in value of rax. If there is use of eax/rax in some instruction, then existing logic should bail out.

Also, I don't understand how implicit-def for $rax is useful in this case?

Reply:
> The implicit operands could be anything, the extra operand could be an arbitrary def of an unrelated register.
 Ok

Reply:
> I'm assuming this pattern appeared as a result of coalescing SUBREG_TO_REG. We should really try to remove it and require emitting instructions that explicitly define the used register parts. We have to go out our way to maintain the liveness it introduced in the producer instructions, and it introduces a lot of complexity in downstream optimizations (like this case). But that will be a big project, particularly for x86

I have observe this pattern after Register Allocation. The live intervals got splitted and a lot copy instructions were generated. Later copy instructions converted to mov32r0 instruction.